graphql-ballerina/query_depth_validator.bal

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+class QueryDepthValidator{
+    private parser:DocumentNode documentNode;
+    private int queryDepth = 0;
+    private int maxQueryDepth = 0;
+    private int queryDepthLimit;
+    private ErrorDetail[] errors;
+
+    public isolated function init(parser:DocumentNode documentNode, int queryDepthLimit){
+        self.documentNode = documentNode;
+        self.queryDepthLimit = queryDepthLimit;
+        self.errors = [];
+    }
+
+    public isolated function validate() returns ErrorDetail[]? {
+        self.visitDocument(self.documentNode);
+        if (self.errors.length() > 0) {
+            return self.errors;
+        }
+    }
+
+    public isolated function visitDocument(parser:DocumentNode documentNode) {
+        parser:OperationNode[] operations = documentNode.getOperations();
+        foreach parser:OperationNode operationNode in operations {
+            self.visitOperation(operationNode);
+            if (self.maxQueryDepth > self.queryDepthLimit) {
+                if (operationNode.getName() != parser:ANONYMOUS_OPERATION) {
+                    string message = string
+                    `Query has depth of ${self.maxQueryDepth}, which exceeds max depth of ${self.queryDepthLimit} on ${
+                        operationNode.getName()}`;
+                    self.errors.push(getErrorDetailRecord(message, operationNode.getLocation()));
+                } else {
+                    string message = string
+                    `Query has depth of ${self.maxQueryDepth}, which exceeds max depth of ${self.queryDepthLimit}`;
+                    self.errors.push(getErrorDetailRecord(message, operationNode.getLocation()));
+                }
+            }
+            self.queryDepth = 0;
+            self.maxQueryDepth = 0;
+        }
+    }
+
+    public isolated function visitOperation(parser:OperationNode operationNode) {
```

Comment:
Shouldn't `visitOperation` increase the `queryDepth` by 1?

Reply:
Operation must have a set of selection otherwise Parser will make sure to fail it. ```queryDepth``` is increased when visiting to that selections.

Reply:
I get that. But since we are starting the `queryDepth` from `0`, shouldn't we increase it by `1` when we visit an operation? Wouldn't this cause incorrect counts?